WIP: Adds transformer pointer generator.

Adamits opened the pull request:
- Adds a pointer generator with a transformer encoder-decoder.
- Introduces a WIP TransformerDecoder that sets a hook to override the torch implementations default behavior of not returning the MHA weights.
- Uses the MHA weights to get pointer probabilities
- Currently does NOT handle features. I believe this will required a second multiheaded attention from each decoder step to the encoded source tokens.  I will note that this may require subclassing the torch `TransformerDecoder` to implement, which will make my work with the hook/`patch_attention` unneeded (we could just hardcode the behavior directly into the subclassed decoder). I have not thought about this much yet though.

This is a WIP, but has a working implementation as far as I can tell, for a transformer encoder-decoder when there are not features. teacher-forced training works by getting all pointer and generator distributions for each target, and then combining them in one go. Greedy decoding for inference works more like the LSTM, where we first decode a single target, then combine pointer and generator probabilities, and then iteratively do it again until we hit EOS or `max_target_length`.

Looking for any obvious mistakes or feedback on the general structure.

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -73,6 +73,26 @@ def forward(
+    """Tracks an attention output during the forward pass.
+    
+    This object can be passed into a hook to the Transformer forward pass
+    in order to modify it's behavior such that certaina ttention weights are 
```
typos: `it's` and `certaina ttention`

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -73,6 +73,26 @@ def forward(
+    def __call__(self, module, module_in, module_out):
+        """Stores the second return argument of `module`.
+        
+        This is intended to be called on a multiehaded attention, which returns
```
typo: `multihaded`

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -73,6 +73,26 @@ def forward(
+        self.outputs.append(module_out[1])
+
+    def clear(self):
+        self.outputs = []
```
Why not just `self.outputs.clear()`? Probably slightly cheaper, certainly clearer to me.

kylebgorman commented on yoyodyne/models/pointer_generator.py:
```diff
@@ -444,3 +447,238 @@ def _reshape_hiddens(
+            context, decoder_output, tgt_embeddings
+        )
+        ptr_scores = (1 - gen_probs) * ptr_probs
+        gen_scores = gen_probs * output_probs
```
consider using `.mul_` here

Adamits replied:
Is it faster?

Adamits replied:
Or are you suggesting because it is in-place?

kylebgorman replied:
Yeah, I am suggesting in-place because it shouddl be faster (clearer too maybe?). the way you wrote it:

* creates a new tensor with (1 - gen_probs) * ptr_probs in it
* assigns it to `ptr_scores`, which triggers atomic increments/decremetns in the ref counter. since the ref count for the older data that `ptr_probs` pointed to should now be zero, it will be marked for (but not yet undergo) GC on the next cycle
* ultimately the GC will also delete the old ptr_probs data

Only the tensor multiplication is necessary if you use the in-place one.

Adamits replied:
Ok. I just need to find the op for a vector * scalar. I get an error with `_mul` because they are not the same shape.

Adamits replied:
I am wrong, I just was calling mul_ on the vector first.

Adamits replied:
These in-place operations seem to break backprop, though I do not understand the error message.

Adamits replied:
I assume its b/c we are overwriting something on those variables that needs to be stored in the CG for backprop, and no longer is?

kylebgorman replied:
FWIW we have used these elsewhere in similar contexts, I believe, and no related bugs yet. If this really does break it I wonder if we're doing "silly scoping" somewhere---i.e., something isn't being passed in or something like that.

Adamits replied:
> something isn't being passed in or something like that.

Can you explain more here? I can look into this a bit harder if it is a symptom of another issue.

kylebgorman replied:
For instance maybe we wrote something that is conceptually similar to:

```python
def foo(bar):
   return baz + bar   # where is baz from? 
```

Barring that I can't imagine how this would influence the computation graph.

kylebgorman replied:
You can always just leave this as a `# TODO: consider in-place operations; using .mul_ here seems to cause issues.` or something like that.

Adamits replied:
Ok, I will do that for now.

bonham79 replied:
pytorch is annoyingly inconsistent with in place operation and maintaining position in the computation graph (also some weird stuff regarding operation managed by ptl AMP for 16 vs 32 bit precision). Safer is usually to just cast to a new tensor (pytorch is usually reliable with only duplicating data when needed).

kylebgorman commented on yoyodyne/models/pointer_generator.py:
```diff
@@ -444,3 +447,238 @@ def _reshape_hiddens(
+        )
+        ptr_scores = (1 - gen_probs) * ptr_probs
+        gen_scores = gen_probs * output_probs
+        return gen_scores + ptr_scores
```
consider using `.add_` here

kylebgorman replied:
Same deal, this is an optimization to avoid thrashing memory/ref counting/GC.

kylebgorman commented on yoyodyne/models/pointer_generator.py:
```diff
@@ -444,3 +447,238 @@ def _reshape_hiddens(
+        return output
+    
+    @property
+    def name(self) -> str:
```
This can be two words, so "pointer-generator transformer" seems fine. Also, you should change the pointer generator LSTM's `name` property to return `pointer-generator LSTM` then I think.

kylebgorman reviewed (commented):
Other than style stuff (I think you know what most of this is, I can call out particular things later) and some notes below, good idea.

I sort of forgot that one could have a true pointer-generator transformer seq2seq model separate from what kind of encoders you have...

Adamits commented:
Thought a bit about adding features to this. I think the two clearest options are:

1. *Best analog to our LSTM implementation* Modify the `TransformerDecoderLayer`s to have a second multiheaded attention that attends to the features. Concatenate the result of this to the lemma attention. Note that we then need to choose a number of attention heads for the features attention, and need to make our model bigger to accomodate this concatenation (recall the transformer represents the "context" vector as the concatenation of all attention ehad results). 
2. Add the features to the input sequence (as in the regular transformer encoder decoder), and mask those positions in the (averaged) multiheaded attention that is used to compute pointer probabilities.

Do you have any thoughts/preference?

kylebgorman commented:
The first one seems more obvious to me. You could then just make it so you have `--source_attention_heads` and `--features_attention_heads` (and get rid of the old `--attention_heads`). The other one seems fine too though honestly. Maybe @bonham79 has an opinion.

Adamits commented:
New commit obviously needs lots of cleanup but:

- This implements 1) from my previous comment
- To achieve this, I subclassed `TransformerDecoderLayer` and `TransformerDecoder`. The layer subclass has a second multiheaded_attention that attends just to the features.  
- In order to combine feature and source attention without breaking the feedforwad layer, nor the residual connection, I have added two linear layers: 1 to map symbol attention to half of the hidden_size, and one to map feature attention to half of the hidden size, and then these are concatenated.

Right now this is hacky in the sense that I probably need to enforce some stuff in the embedding/hidden size configurations, and the code to choose whether to use features or not is sloppy and redundant. Obviously I also need to cleanup documentation. I ran this with a sanity checking script below, and it was converging pretty quickly.

```
#!/bin/bash

LANGUAGE=english

readonly ROOT="${USER}/nlp-projects/morphology"
readonly DATA="${ROOT}/conll2017/all/task1/"
readonly TRAIN="${DATA}/${LANGUAGE}-train-high"
readonly DEV="${DATA}/${LANGUAGE}-dev"

# for ARCH in attentive_lstm lstm pointer_generator_lstm pointer_generator_transformer transformer transducer; do
ARCH=pointer_generator_transformer
RESULTS_PATH="./results/${LANGUAGE}/${ARCH}"

yoyodyne-train \
    --arch "${ARCH}" \
    --encoder_layers 2 \
    --decoder_layers 2 \
    --embedding_size 64 \
    --hidden_size 64 \
    --target_col 2 \
    --features_col 3 \
    --features_encoder_arch transformer \
    --accelerator cpu \
    --experiment "${LANGUAGE}" \
    --train "${TRAIN}" \
    --val "${DEV}" \
    --label_smoothing 0.1 \
    --batch_size 32 \
    --max_epochs 4 \
    --save_top_k 1 \
    --check_val_every_n_epoch 1 \
    --model_dir "${RESULTS_PATH}" ;
```

kylebgorman commented:
Cool. I'm mostly following along. Let me know when you're ready for review.

Adamits commented on yoyodyne/models/__init__.py:
```diff
@@ -26,6 +29,7 @@ def get_model_cls(arch: str) -> BaseEncoderDecoder:
         "attentive_lstm": AttentiveLSTMEncoderDecoder,
         "lstm": LSTMEncoderDecoder,
         "pointer_generator_lstm": PointerGeneratorLSTMEncoderDecoder,
+        "pointer_generator_transformer": PointerGeneratorTransformerEncoderDecoder,
```
Flake8 complains about this. If I put the model on a new line, black complains. @kylebgorman do you know the solution here?

kylebgorman replied:
Just use a `  # noqa` on the offending lines.

Adamits commented:
Ready for review! Notice my inline comment about why CI is failing.

kylebgorman commented on yoyodyne/defaults.py:
```diff
@@ -14,6 +14,7 @@
 # Architecture arguments.
 ARCH = "attentive_lstm"
 ATTENTION_HEADS = 4
```
I think for clarity I'd prefer `SOURCE_ATTENTION_HEADS` and `FEATURES_ATTENTION_HEADS` (note for consistency I am always using "features" and not "feature"). And same for the flags. We do this with encoder modules too---distinguish between source and features even though sometimes source includes features because concatenation...

Adamits replied:
It may be that `SOURCE_ATTENTION_HEADS` is the best name, but wanted to clarify that we reuse this variable for self attention in both the encoder and decoder, as well as for the "cross-attention" (can't remember the canonical name for this) from encoder to decoder. I think using the same number for all of these is standard. SOURCE could imply that it is specific to the encoder, or just the "cross" attention?

kylebgorman replied:
I see and agree. Not a bad idea to comment that somewhere in the source (maybe not here though).

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -73,6 +73,42 @@ def forward(
+
+    def __init__(self):
+        """Initializes an AttentionOutput."""
+        self.outputs = []
```
maybe declare this type for the class

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -184,6 +220,242 @@ def name(self) -> str:
+
+    def forward(
+        self,
+        tgt: torch.Tensor,
```
For consistency with what we do elsewhere I suggest `target_*` rather than `tgt_`.

Adamits replied:
I think I had forgotten why I had `tgt`, and then I must not have tested before changing it here :( sorry! (see comment about this error).

kylebgorman commented on yoyodyne/ptr_gen_trm_stesp.log:
remove this

Adamits replied:
Oops!

kylebgorman reviewed (commented):
Added some comments. A bunch of "seperate" typos (should be "separate"). One of them looked like it might be not-just-no-op actually.

Adamits commented:
Not quite sure what happened. I had pulled your commit, but still somehow needed to then merge either my next commit or yours back in?

Anyway, I think this is in a good state now given your notes. I did do the `ATTENTION_HEADS -> SOURCE_ATTENTION_HEADS` change. It seems ok for now?

kylebgorman commented:
I made some ninja style edits so you had to do a merge those back in. Will
take a final look soon.

On Fri, Aug 18, 2023 at 9:56 PM Adam ***@***.***> wrote:

> Not quite sure what happened. I had pulled your commit, but still somehow
> needed to then merge either my next commit or yours back in?
>
> Anyway, I think this is in a good state now given your notes. I did do the ATTENTION_HEADS
> -> SOURCE_ATTENTION_HEADS change. It seems ok for now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CUNY-CL/yoyodyne/pull/135#issuecomment-1684472497>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OISUTIMYSG3KLVUXIDXV7QIDANCNFSM6AAAAAA3TFSGEU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bonham79 commented:
> Thought a bit about adding features to this. I think the two clearest options are:
> 
>     1. _Best analog to our LSTM implementation_ Modify the `TransformerDecoderLayer`s to have a second multiheaded attention that attends to the features. Concatenate the result of this to the lemma attention. Note that we then need to choose a number of attention heads for the features attention, and need to make our model bigger to accomodate this concatenation (recall the transformer represents the "context" vector as the concatenation of all attention ehad results).
> 
>     2. Add the features to the input sequence (as in the regular transformer encoder decoder), and mask those positions in the (averaged) multiheaded attention that is used to compute pointer probabilities.
> 

Late on the draw but  simply adding features to input seems the most unobtrusive result. (In my mind it works similar to positional encoding, so don't see why similar can't be done here).

Adamits commented:
> > Thought a bit about adding features to this. I think the two clearest options are:
> > ```
> > 1. _Best analog to our LSTM implementation_ Modify the `TransformerDecoderLayer`s to have a second multiheaded attention that attends to the features. Concatenate the result of this to the lemma attention. Note that we then need to choose a number of attention heads for the features attention, and need to make our model bigger to accomodate this concatenation (recall the transformer represents the "context" vector as the concatenation of all attention ehad results).
> > 
> > 2. Add the features to the input sequence (as in the regular transformer encoder decoder), and mask those positions in the (averaged) multiheaded attention that is used to compute pointer probabilities.
> > ```
> 
> Late on the draw but simply adding features to input seems the most unobtrusive result. (In my mind it works similar to positional encoding, so don't see why similar can't be done here).

Sure, we should add this option for both Transformer and LSTM. I was hypothesizing that its probably trivial to learn not to 'point' to features. I just talked to a lab mate who uses someone else's implementation of transformer pointer-generator apparently, and indeed thinks concatenating the features works fine, and it doesn't really seem to matter if you mask or not.

kylebgorman commented:
> Sure, we should add this option for both Transformer and LSTM. I was hypothesizing that its probably trivial to learn not to 'point' to features. I just talked to a lab mate who uses someone else's implementation of transformer pointer-generator apparently, and indeed thinks concatenating the features works fine, and it doesn't really seem to matter if you mask or not.

Probably true with enough (for some value of "enough") data but seems like a waste of parameters.

We don't necessarily need to support concatenation at all since the user can always concatenate TSV columns, but we do need to support separate encoders since there's no UNIX trick for that.

Adamits commented:
Does this PR require updates from me?

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
+                target_key_padding_mask=target_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+            )
+        if self.norm is not None:
```
possibly dumb question, does this need to be nested inside the for-loop that begins at line 268?

Adamits replied:
Oh i will double check.

Adamits replied:
Adding a norm after each layer looks right.

Adamits replied:
Actually, I am not sure I see this in the original paper, but it is there in the torch implementation.

Adamits replied:
Ah, it is optional and we do not set it. I don't see a citation in their code for when you'd want to set it, but we should keep it here in case that option is set.

kylebgorman replied:
Okay but maybe we should make a note to try layer norm. I thought that was the standard for transformer encoder layers.

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
+            # is_causal=is_causal,
+            need_weights=False,
+        )[0]
+        return self.dropout2(x)
```
Another dumb question: where is the `dropout2` object or method implemented? Did I just miss it?

Adamits replied:
I believe this is in the torch `TransformerDecoderLayer` that I inherit from.

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
+        return output
+
+
+class TransformerDecoderLayerSeperateFeatures(nn.TransformerDecoderLayer):
```
typo: `Seperate` should be `Separate`.

kylebgorman commented on yoyodyne/models/modules/transformer.py:
```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
+            **factory_kwargs,
+        )
+        self.features_linear = nn.Linear(
+            kwargs["d_model"],  # TODO: Seperate feature embedding size?
```
typo: "Seperate" for "separate"

kylebgorman reviewed (commented):
Sorry @Adamits I thought this wasn't working yet. (Do we have to migrate to PyTorch 2 also to get this working?) 

I added two dumb questions, and a few typos. 

I have a smarter question. Right now you're using a divide-by-two trick, which as you note won't work if the dimensionality is odd. Why divide at all? I realize this will give us a bigger actual feedforward layer but I think that's what most people would expect to happen. 

@bonham79 any thoughts on this PR in general?

Adamits commented:
> Sorry @Adamits I thought this wasn't working yet. (Do we have to migrate to PyTorch 2 also to get this working?)

No it should work. I have comments about flags we might need when we go to torch 2.0. This is because I could only find the references to certain torch things (e.g. _sa_block) in torch 2.0, and some of the args I passed were throwing errors in my <2.0 torch.

> I added two dumb questions, and a few typos.
> 
> I have a smarter question. Right now you're using a divide-by-two trick, which as you note won't work if the dimensionality is odd. Why divide at all? I realize this will give us a bigger actual feedforward layer but I think that's what most people would expect to happen.

This is because I want a separate symbols and features multiheaded attention, so I can query only the symbols. But to run the Transformer it expects representations of size `h` in the feedforward block. Since the torch implementation iirc computes the mh attention sizes from the number of attention heads, and feedforward size, we need to half the size of each since for each attention head, we know have TWO representations (one for symbols and one for features), and then concatenate them before passing them through the feedforward layer. I could instead keep the dimensions the same and average them, but concat seems better to me.

> @bonham79 any thoughts on this PR in general?

kylebgorman commented:
> > Sorry @Adamits I thought this wasn't working yet. (Do we have to migrate to PyTorch 2 also to get this working?)
> 
> No it should work. I have comments about flags we might need when we go to torch 2.0. This is because I could only find the references to certain torch things (e.g. _sa_block) in torch 2.0, and some of the args I passed were throwing errors in my <2.0 torch.
> 
> > I added two dumb questions, and a few typos.
> > I have a smarter question. Right now you're using a divide-by-two trick, which as you note won't work if the dimensionality is odd. Why divide at all? I realize this will give us a bigger actual feedforward layer but I think that's what most people would expect to happen.
> 
> This is because I want a separate symbols and features multiheaded attention, so I can query only the symbols. But to run the Transformer it expects representations of size `h` in the feedforward block. Since the torch implementation iirc computes the mh attention sizes from the number of attention heads, and feedforward size, we need to half the size of each since for each attention head, we know have TWO representations (one for symbols and one for features), and then concatenate them before passing them through the feedforward layer. I could instead keep the dimensions the same and average them, but concat seems better to me.

Concat is fine. Why not act as if when people request `h` that it was actually `2h` and use `h` of them for source symbols and `h` of them for features?

Adamits commented:
> > > Sorry @Adamits I thought this wasn't working yet. (Do we have to migrate to PyTorch 2 also to get this working?)
> > 
> > 
> > No it should work. I have comments about flags we might need when we go to torch 2.0. This is because I could only find the references to certain torch things (e.g. _sa_block) in torch 2.0, and some of the args I passed were throwing errors in my <2.0 torch.
> > > I added two dumb questions, and a few typos.
> > > I have a smarter question. Right now you're using a divide-by-two trick, which as you note won't work if the dimensionality is odd. Why divide at all? I realize this will give us a bigger actual feedforward layer but I think that's what most people would expect to happen.
> > 
> > 
> > This is because I want a separate symbols and features multiheaded attention, so I can query only the symbols. But to run the Transformer it expects representations of size `h` in the feedforward block. Since the torch implementation iirc computes the mh attention sizes from the number of attention heads, and feedforward size, we need to half the size of each since for each attention head, we know have TWO representations (one for symbols and one for features), and then concatenate them before passing them through the feedforward layer. I could instead keep the dimensions the same and average them, but concat seems better to me.
> 
> Concat is fine. Why not act as if when people request `h` that it was actually `2h` and use `h` of them for source symbols and `h` of them for features?

That is probably fine too, but " act as if when people request h" might require some fiddling with some deeper abstractions in torch iirc. I can't remember clearly so I should look into this either way.

Adamits commented:
Hey Kyle,

Oh good catch. I reproduced the `target_mask` for both cases. Not sure what I overlooked before,but hopefully something simple. I will check it out soon.

Adamits commented:
> Hey Kyle,
> 
> Oh good catch. I reproduced the `target_mask` for both cases. Not sure what I overlooked before,but hopefully something simple. I will check it out soon.

I think its due to a mismatch between our interface (`target_mask`) and the torch 1.13.1 interface (`tgt_mask`). But I also thought this particular scripts should hit my interface anyway, so need to investigate further.

Adamits commented:
Ok for Error 2. from your comment, I found that I was setting the output size of the feed forward layers I added for compressing feature/symbols before concatenation. I have now fixed it.

I must have had the settings wrong in my tests, thank you for testing!!!

I will push what I have for the fixes now, but this PR still needs some smoothing out (re: `tgt` v.s. `target` at the very least).

kylebgorman commented:
Would it be helpful for me to test again at this point? Or do you want more time with it?

Adamits commented:
If you could test again, that would be awesome :)

On Thu, Sep 21, 2023 at 10:37 AM Kyle Gorman ***@***.***>
wrote:

> Would it be helpful for me to test again at this point? Or do you want
> more time with it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CUNY-CL/yoyodyne/pull/135#issuecomment-1729928699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIQBSXHGEM2WCYJQFVCPBDX3RUMXANCNFSM6AAAAAA3TFSGEU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kylebgorman reviewed (commented):
Thanks Adam.

* Features-less (Maori data) works on GPU as of this commit and achieves a decent accuracy
* Features-full (Polish data, I am using the linear feature encoder) also works on GPU, but accuracy is awful: https://wandb.ai/hyperparam-sensitivity/pol/runs/nvtn0q1o?workspace=user-kylebgorman

A note for later. Validation (at least for Polish) takes a rather long time, the data loader visually appears to have freezed if you're watching the progress bars and it accounts for a lot of the wall clock time. I wonder if the batches are just too big; recall we are using the trick where my already big batch sizes--400?--are twice as large during validation, when I'm not collecting gradients, as during training. Maybe that was not a good heuristic after all.

Adamits commented:
> Thanks Adam.
> 
> * Features-less (Maori data) works on GPU as of this commit and achieves a decent accuracy
> * Features-full (Polish data, I am using the linear feature encoder) also works on GPU, but accuracy is awful: https://wandb.ai/hyperparam-sensitivity/pol/runs/nvtn0q1o?workspace=user-kylebgorman

Thanks so much for the testing.

I wonder if my good results for this were inadvertently calling it without features. I will be debugging today. 

> A note for later. Validation (at least for Polish) takes a rather long time, the data loader visually appears to have freezed if you're watching the progress bars and it accounts for a lot of the wall clock time. I wonder if the batches are just too big; recall we are using the trick where my already big batch sizes--400?--are twice as large during validation, when I'm not collecting gradients, as during training. Maybe that was not a good heuristic after all.

Does it take significantly longer than the vanilla transformer? I wonder how much bigger the model is. I recall the pointer-generator is a lot slower than the vanilla LSTM with the same parameters. I can play with hyperparameters a bit to see if batch size should be smaller anyway. But yeah maybe we want to have an eval batch size option (we removed it right?), and just default to twice the train batch size with a message that we log saying as much.

kylebgorman commented:
> > A note for later. Validation (at least for Polish) takes a rather long time, the data loader visually appears to have freezed if you're watching the progress bars and it accounts for a lot of the wall clock time. I wonder if the batches are just too big; recall we are using the trick where my already big batch sizes--400?--are twice as large during validation, when I'm not collecting gradients, as during training. Maybe that was not a good heuristic after all.
> 
> Does it take significantly longer than the vanilla transformer? I wonder how much bigger the model is. 

> I recall the pointer-generator is a lot slower than the vanilla LSTM with the same parameters.

I do'nt have this experience: the ptr-generator is about as performant as the vanilla LSTM and converges muchhhh faster.

> I can play with hyperparameters a bit to see if batch size should be smaller anyway. But yeah maybe we want to have an eval batch size option (we removed it right?)...

If I'm right and the batch size is too large, I disagree, we should just use the same one for train and validation and get both from `--batch_size`. We already have too many options.

Adamits commented:
> I do'nt have this experience: the ptr-generator is about as performant as the vanilla LSTM and converges muchhhh faster.

Sorry I meant that one step of inference takes much longer. Agree that it converges in far fewer steps.

> If I'm right and the batch size is too large, I disagree, we should just use the same one for train and validation and get both from --batch_size. We already have too many options.

Ok, I think that's fine! Tbf I think its pretty normal to give an option for different train/inference batch sizes in these libraries, but I agree we have too many options so your recommendation is the best choice!

bonham79 commented:
@Adamits Do any of the transformer modules use `batch_norm` or is it all `layer_norm`. If the former, we probably should make a flag for `eval_batch_size`

Adamits commented:
@bonham79 Should just be layer norm.

kylebgorman commented:
> Hey Kyle,
> 
> I think this may be a hyperparameters issue and not a bug. 

Nice. I can confirm this evening. I mostly borrowed hyperparameters from the Wu et al. paper even though they weren't looking at a pointer-generator model.

kylebgorman commented:
> @bonham79 Should just be layer norm.

Not directly related to this PR but I would like to confirm that we're using this like I assumed we are.

kylebgorman commented:
Yes, this smaller model works just fine, thanks Adam. Should I merge now?

Adamits commented:
> Yes, this smaller model works just fine, thanks Adam. Should I merge now?

Great! I will do a final pass to make sure things make sense here. I will hopefully get to it in a few hours, then I will ping you here.

kylebgorman commented:
> > Not directly related to this PR but I would like to confirm that we're using this like I assumed we are.
> 
> Ok, should we talk about this? I have never really thought about batch norm, but definitely did not explicitly implement anything to use it. In the transformer, I implemented it to use layer norm, as is typical, and we have the option of doing the norm before or after the FF layer (inside the residual block or after).

This is all I wanted to check, that encoder and decoder have some form of layer norm: 

* [In encoder](https://github.com/Adamits/yoyodyne/blob/807874bc20c51c798e31077b96f2e1debcbfda7d/yoyodyne/models/modules/transformer.py#L208) 
* [In decoder](https://github.com/Adamits/yoyodyne/blob/807874bc20c51c798e31077b96f2e1debcbfda7d/yoyodyne/models/modules/transformer.py#L577) 

LGTM.

bonham79 commented:
Alrighty, not a big deal then. Just wanted to check

Adamits commented:
Ok made a last pass and added a `TODO`. I think this can be merged now though.

kylebgorman reviewed (approved):
LGTM, merging now.
